docs/compilers/terrapin.md

```diff
@@ -0,0 +1,125 @@
+
+1. Execute `dotnet-roslyn-manifest-generator` on the DLL + PDB conversation. 
+1. Setup the environment for validation:
+    1. Transfer execution to the appropriate operating system listed in the manifest. All future actions are expected to run on this machine.
+    1. Install the correct .NET Runtime on the target machine (if necessary)
+1. Download the artifacts listed in the artifact manifest file as well as the specified version of `dotnet-roslyn-validator`. 
+1. Execute `dotnet-roslyn-validator` providing the DLL, PDB and the directory where the artifacts were downloaded too.
+
+The Terrapin service will download the artifacts into a directory created by Terrapin. The directory will have two directories to store source files and references respectively: sources and references. The name of the artifacts in those directories will be specified in the manifest file. That will put the burden of avoiding collisions and dealing with file path fixup on the compiler team.
+
+## Artifacts Manifest File
+The goal of `dotnet-roslyn-manifest-generator` is to provide a manifest file to the Terrapin service which lists all of the artifacts necessary for a validation to be completed as well as the environment needed to run the validation. The set of artifacts that are necessary include:
+
+1. The source files used in the compilation. 
+1. The reference DLL used in the compilation. 
+
+At this time the expectation is Terrapin will need to recreate much of the environment used to create the binaries in order to perform provenance validation. This includes having a matching .NET Runtime version and operating system. This is because OS dependencies, like zip compression version, can impact the deterministic output of the compiler. It's possible our experiments in Terrapin will cause us to re-evaluate how these play into determinism but for now the design is to build validation on top of our existing assumptions here. 
+
+To facilitate this the artifact manifest file will include all of the necessary environment information. This will include the OS RID, the .NET Runtime version and whether or a desktop or NET Core runtime was used to run the compiler.
+
+The artifacts manifest file will be a JSON file with the following format:
+
+```json
+{
+    "sources": [
+        { 
+            "uri": "https://github.com/dotnet/roslyn/blob/f5517311e37a3883e09dc6daa5069bbc4f44c882/src/Compilers/Shared/Csc.cs",
+            "destFileName": "c6fea380dac7392b945fa6c63466746f830c5cc3"
+        }
+    ],
+    "references": [
+        {
+            "name": "System.Reflection.Metadata.dll",
+            "mvid": "51ef3ce0-5a89-4325-b094-8a5970acc676",
+            "destFileName": "51ef3ce0-5a89-4325-b094-8a5970acc676",
+        }
+    ],
+    "environment": {
+        "dotnet-roslyn-validator-version": "3.8.0",
+        "os": "windows",
+        "dotnet-runtime": "net5.0.0"
```

Comment:
Include `os-version` as well?

Reply:
Yeah we definitely need to include the OS version. Do you have any estimations of what the best way to encode this would be? Thought possibly RID style.

Reply:
I think we need more than RID, since compression libs can change update to update. 
`"os-name": "windows"` is kind of the RID and  `"os-version": "10.1909"` or something

![image](https://user-images.githubusercontent.com/41759/102143524-eb22dd80-3e18-11eb-9239-9b23cb8e097a.png)

Reply:
Do you have a sense of what properties / values from the .NET APIs we should be encoding here?

Reply:
Thinking about this more... The only OS component that the compiler depends on is the compression library. Rather then dealing with various Linux distro version, perhaps it would be best if .NET Core exposed APIs to get the file version of the compression library. We'd handle .NET Framework separately - depending on whether the compression lib is part of .NET Framework or Windows OS (not sure atm, would need to find out).

Reply:
From a tooling perspective I think just enumerating dependencies individually is best instead of grouping into "OS". There's a reason we depend on that OS in some way, and adding compression library version would be great. It does mean potentially harder to maintain because we might miss a dependency in the future, and we may need to provide a known mapping of library version to OS